openmdao/solvers/linear/direct.py

```diff
@@ -270,6 +270,101 @@ def _linearize(self):
                 except ValueError as err:
                     raise RuntimeError(format_nan_error(system, mtx))
 
+    def _inverse(self):
+        """
+        Return the inverse Jacobian.
+
+        This is only used by the Broyden solver when calculating a full model Jacobian. Since it
+        is only done for a single RHS, no need for LU.
+
+        Returns
+        -------
+        ndarray
+            Inverse Jacobian.
+        """
+        system = self._system
+
+        if self._assembled_jac is not None:
+
+            mtx = self._assembled_jac._int_mtx
+            ranges = self._assembled_jac._view_ranges[system.pathname]
+            matrix = mtx._matrix[ranges[0]:ranges[1], ranges[0]:ranges[1]]
+
+            # Perform dense or sparse lu factorization
+            if isinstance(mtx, DenseMatrix):
+                # Detect singularities and warn user.
+                with warnings.catch_warnings():
+                    if self.options['err_on_singular']:
+                        warnings.simplefilter('error', RuntimeWarning)
+                    try:
+                        inv_jac = scipy.linalg.inv(matrix)
+                    except RuntimeWarning as err:
+                        raise RuntimeError(format_singular_error(err, system, matrix))
+
+                    # NaN in matrix.
+                    except ValueError as err:
+                        raise RuntimeError(format_nan_error(system, matrix))
+
+            elif isinstance(mtx, (CSRMatrix, CSCMatrix)):
+                try:
+                    inv_jac = scipy.sparse.linalg.inv(matrix)
+                except RuntimeError as err:
+                    if 'exactly singular' in str(err):
+                        raise RuntimeError(format_singular_csc_error(system, matrix))
+                    else:
+                        reraise(*sys.exc_info())
+            else:
+                raise RuntimeError("Direct solver not implemented for matrix type %s"
+                                   " in system '%s'." % (type(mtx), system.pathname))
+
+        else:
```

Comment:
I've:
1. Refactored the direct solver stuff a bit and moved the mvp jac generation into it's own method to get rid of the duplication.
2. Updated Broyden to use the full residual vector to evaluate loop convergence.
3. Added a test where Broyden solves a Sellar loop, though I think the bulk of the solution comes about because running the workflow during iteration mimics gauss seidel.